src/statsd_client.cpp

```diff
@@ -32,217 +37,150 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #include <compat.h>
 #include <netbase.h>
-#include <random.h>
 #include <util/system.h>
 
 #include <cmath>
 #include <cstdio>
+#include <random>
 
 std::unique_ptr<statsd::StatsdClient> g_stats_client;
 
 namespace statsd {
-
-inline bool fequal(float a, float b)
+bool StatsdClient::ShouldSend(float sample_rate)
 {
-    const float epsilon = 0.0001;
-    return ( fabs(a - b) < epsilon );
+    LOCK(cs);
+    sample_rate = std::clamp(sample_rate, 0.f, 1.f);
+    constexpr float EPSILON{0.0001f};
+    return (std::fabs(sample_rate - 1.f) < EPSILON ||
+            sample_rate > std::uniform_real_distribution<float>(0.f, 1.f)(insecure_rand));
 }
 
-thread_local FastRandomContext insecure_rand;
-
-inline bool should_send(float sample_rate)
+StatsdClient::StatsdClient(const std::string& host, const std::string& nodename, uint16_t port, const std::string& ns,
+                           bool enabled)
+    : m_port{port}, m_host{host}, m_nodename{nodename}, m_ns{ns}
 {
-    if ( fequal(sample_rate, 1.0) )
-    {
-        return true;
+    if (!enabled) {
+        LogPrintf("Transmitting stats are disabled, will not init StatsdClient\n");
+        return;
     }
 
-    float p = float(insecure_rand(std::numeric_limits<uint32_t>::max())) / float(std::numeric_limits<uint32_t>::max());
-    return sample_rate > p;
-}
-
-struct _StatsdClientData {
-    SOCKET  sock;
-    struct  sockaddr_in server;
-
-    std::string  ns;
-    std::string  host;
-    std::string  nodename;
-    short   port;
-    bool    init;
-
-    char    errmsg[1024];
-};
-
-StatsdClient::StatsdClient(const std::string& host, int port, const std::string& ns) :
-    d(std::make_unique<_StatsdClientData>())
-{
-    d->sock = INVALID_SOCKET;
-    config(host, port, ns);
-}
-
-StatsdClient::~StatsdClient()
-{
-    // close socket
-    CloseSocket(d->sock);
-}
-
-void StatsdClient::config(const std::string& host, int port, const std::string& ns)
-{
-    d->ns = ns;
-    d->host = host;
-    d->port = port;
-    d->init = false;
-    CloseSocket(d->sock);
-}
-
-int StatsdClient::init()
-{
-    static bool fEnabled = gArgs.GetBoolArg("-statsenabled", DEFAULT_STATSD_ENABLE);
-    if (!fEnabled) return -3;
-
-    if ( d->init ) return 0;
-
-    config(gArgs.GetArg("-statshost", DEFAULT_STATSD_HOST), gArgs.GetArg("-statsport", DEFAULT_STATSD_PORT), gArgs.GetArg("-statsns", DEFAULT_STATSD_NAMESPACE));
-
-    d->sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
-    if ( d->sock == INVALID_SOCKET ) {
-        snprintf(d->errmsg, sizeof(d->errmsg), "could not create socket, err=%m");
-        return -1;
+    CNetAddr netaddr;
+    if (!LookupHost(m_host, netaddr, /*fAllowLookup=*/false) || !(netaddr.IsIPv4() || netaddr.IsIPv6())) {
+        LogPrintf("Unable to lookup host %s, cannot init StatsdClient\n", m_host);
```

Comment:
nit: add "ERROR" to text or something similar?

Reply:
Resolved in latest push!